src/rosdiscover/observer/ros1.py

```diff
@@ -0,0 +1,304 @@
+# -*- coding: utf-8 -*-
+import re
+from typing import Dict, Optional
+
+from loguru import logger
+
+from .observer import Observer
+from ..core import Topic
+from ..interpreter import NodeContext, ParameterServer
+
+_GOAL = re.compile(r'(.*)/goal')
+_GOAL_FMT = re.compile(r'(.*)Goal')
+_CANCEL = re.compile(r'(.*)/cancel')
+_STATUS = re.compile(r'(.*)/status')
+_FEEDBACK = re.compile(r'(.*)/feedback')
+_FEEDBACK_FMT = re.compile(r'(.*)Feedback')
+_RESULT = re.compile(r'(.*)/result')
+_RESULT_FMT = re.compile(r'(.*)Result')
+
+_NODES_TO_FILTER_OUT = {'/rosout'}
+_TOPICS_TO_FILTER_OUT = {'/rosout', '/rosout_agg'}
+_SERVICES_TO_FILTER_OUT = {'set_logger_level', 'get_loggers'}
+
+
+class ActionCandidate:
+    goal: Optional[Topic]
+    cancel: Optional[Topic]
+    status: Optional[Topic]
+    feedback: Optional[Topic]
+    result: Optional[Topic]
+
+    def __init__(self,
+                 action_name: str,
+                 is_server: bool):
+        self._action_name = action_name
+        self._server = is_server
+        self.goal = None
+        self.cancel = None
+        self.status = None
+        self.feedback = None
+        self.result = None
+
+    @property
+    def name(self):
+        return self._action_name
+
+    @property
+    def fmt(self):
+        assert self.is_complete()
+        return _GOAL_FMT.match(self.goal.format).group(1)
+
+    def is_complete(self):
+        """Check if all information is there for an action.
+
+        Returns
+        ------_
+        bool
+            Whether the action is complete
+        """
+        return self.goal is not None and self.cancel is not None and self.status is not None \
```

Comment:
This is a little difficult to parse. Consider splitting up.

Reply:
Done